infra/build/functions/update_build_status.py

```diff
@@ -24,71 +24,128 @@
+      break
 
+  return history, last_build_status
 
+
+# pylint: disable=too-many-locals
 def update_build_status(build_tag, status_filename):
   """Update build statuses."""
+  build_history = []
+  last_successful_build = []
   statuses = {}
   successes = []
   failures = []
+  not_yet_built = []
   for project_build in BuildsHistory.query(
-      BuildsHistory.build_tag == build_tag):
-    last_build = get_last_build(project_build.build_ids)
-    if not last_build:
-      logging.error('Failed to get last build for project %s',
-                    project_build.project)
+      BuildsHistory.build_tag == build_tag).order('project'):
+
+    history, last_build_status = get_build_history(project_build.build_ids)
+
+    if not last_build_status:
+      not_yet_built.append({'name': project_build.project, 'history': history})
       continue
 
-    if last_build['status'] == 'SUCCESS':
+    if last_build_status == 'SUCCESS':
       statuses[project_build.project] = True
-      successes.append({
-          'name': project_build.project,
-          'build_id': last_build['id'],
-          'finish_time': last_build['finishTime'],
-          'success': True,
-      })
+      successes.append({'name': project_build.project, 'history': history})
     else:
       statuses[project_build.project] = False
-      failures.append({
-          'name': project_build.project,
-          'build_id': last_build['id'],
-          'finish_time': last_build['finishTime'],
-          'success': False,
-      })
-
-  builds_status.upload_status(successes, failures, status_filename)
+      failures.append({'name': project_build.project, 'history': history})
+
+  for project in failures:
+    build_history.append(project)
+    last_success = ndb.Key(LastSuccessfulBuild,
+                           project.name + '-' + build_tag).get()
+    if not last_success:
+      continue
+    last_successful_build.append({
+        'name': project.name,
+        'build_id': last_success.build_id,
+        'finish_time': last_success.finish_time
+    })
+
+  for project in not_yet_built:
+    build_history.append(project)
+
+  for project in successes:
+    last_success = ndb.Key(LastSuccessfulBuild,
+                           project.name + '-' + build_tag).get()
+    if not last_success:
+      last_success = LastSuccessfulBuild(
+          id=project.name + '-' + build_tag,
+          project=project.name,
+          build_id=project.history[0].build_id,
+          finish_time=project.history[0].finish_time)
+    else:
+      last_success.build_id = project.history[0].build_id
+      last_success.finish_time = project.history[0].finish_time
+
+    last_success.put()
+    build_history.append(project)
+    last_successful_build.append({
+        'name': project.name,
+        'build_id': last_success.build_id,
+        'finish_time': last_success.finish_time
+    })
+
+  data = {
+      'last_successful_build': last_successful_build,
+      'build_history': build_history
```

Comment:
let's call this field 'projects'. it's a little strange that the structure has a top level build_history, and then each value also has a 'history' field. 

Let's also move last successful build into the same project value to make everything more logical. i.e. the data should look like: 
 
```
{
   "projects": [
       {"last_successful_build": ...
         "history": [...]
       },
       ...
   ]
}
```